### PR TITLE
Pruning export

### PR DESCRIPTION
**Pull request by mkaglins**

### CODE REVIEW

File: nncf/dynamic_graph/graph.py [hunk 1]
```diff
@@ -218,13 +218,14 @@ def add_node(self, op_exec_context: OperationExecutionContext, inputs) -> NNCFNo
...
         has_traced_inputs = False
-        for info in op_exec_context.tensor_metas:
+        for i, info in enumerate(op_exec_context.tensor_metas):
             if info is None or info.creator_id is None:
                 continue
             parent = self._node_id_to_key_dict[info.creator_id]
             self._nx_graph.add_edge(parent, node_key)
             has_traced_inputs = True
             self._nx_graph.edges[parent, node_key][NNCFGraph.ACTIVATION_SHAPE_EDGE_ATTR] = info.shape
+            self._nx_graph.edges[parent, node_key]['in_port'] = i
```
Comment (ljaljushkin) [inline nncf/dynamic_graph/graph.py:228]: `in_port` string should be a const variable similar to `NNCFGraph.ACTIVATION_SHAPE_EDGE_ATTR`

File: nncf/pruning/export_helpers.py [hunk 2]
```diff
@@ -0,0 +1,394 @@
...
+
+    def propagate_can_prune_attr_up(self):
+        """
+        Propagating can_prune attribute in reversed topological order.
+        This attribute depends on accept_pruned_input and can_prune attributes of output nodes.
+        Node can_prune is True if all outputs accept_pruned_input is True and all outputs
+        (except convs because conv can be pruned by input and output independently).
+        """
+        for node_name in self.nx_graph.nodes:
+            self.nx_graph.nodes[node_name]['can_prune'] = True
```
Comment (ljaljushkin) [inline nncf/pruning/export_helpers.py:325]: I would use constant variable for this string as well, because it's used in multiple places and there's a chance to introduce  a bug on refactoring

**Comment by AlexKoff88** [comment]
Need to claim results for the CNNs we agreed on.

File: nncf/pruning/export_helpers.py [hunk 3]
```diff
@@ -0,0 +1,394 @@
...
+from nncf.nncf_network import NNCFNetwork
+from nncf.pruning.export_utils import PruningOperationsMetatypeRegistry, identity_mask_propagation, get_input_masks, \
+    fill_input_masks
+from nncf.pruning.utils import get_sources_of_node
+
+PRUNING_OPERATOR_METATYPES = PruningOperationsMetatypeRegistry("operator_metatypes")
+
+
+# pylint: disable=protected-access
+class DefaultMetaOp:
```
Comment (AlexKoff88) [inline nncf/pruning/export_helpers.py:31]: It would be great to have a description of the whole idea of the export algorithm [here](https://github.com/openvinotoolkit/nncf_pytorch/blob/master/docs/compression_algorithms/Pruning.md)

**Comment by mkaglins** [comment]
> Need to claim results for the CNNs we agreed on.

Results for all Resnet models:

| Model name | Requested PR | Actual PR | FLOPs  pruned|
|------------|--------------|-----------|-------|
| resnet18   |      30%     |   15.8%   | 16.9% |
| ressnet34  |      30%     |   21.7%   | 22.1% |
| resnet50   |      30%     |   28.3%   |  30%  |

**Comment by mkaglins** [comment]
@vshampor Please, take a look.

File: nncf/pruning/export_helpers.py (hunk 3 above)
Comment (AlexKoff88) [reply nncf/pruning/export_helpers.py:31]: I see problems with formating when visualizing .md file. Please also consider using header (##) when formatting the text.

File: nncf/pruning/base_algo.py [hunk 4]
```diff
@@ -239,3 +267,76 @@ def statistics(self):
...
+
+        hook_list = [m.register_forward_hook(get_hook(n)) for n, m in model.named_modules()]
+
+        model.do_dummy_forward(force_eval=True)
+
+        for h in hook_list:
+            h.remove()
+
+        total_ops_count = sum(v for v in ops_count_dict.values())
+        return total_ops_count
```
[the remark this answers is not in the document]
Comment (mkaglins) [reply nncf/pruning/base_algo.py]: Functions replaced.

**Comment by vshampor** [comment]
Jenkins please retry a build

File: nncf/pruning/export_helpers.py (hunk 3 above)
Comment (mkaglins) [reply nncf/pruning/export_helpers.py:31]: @AlexKoff88 formatting problems were fixed. Any other comments?

File: docs/compression_algorithms/Pruning.md [hunk 5]
```diff
@@ -83,6 +84,30 @@ Where ![a, k, b](https://microsoft.codecogs.com/gif.latex?a%2C%20k%2C%20b) - par
...
     // "target_scopes": []
-}```
+}
+```
 
 > **NOTE:**  In all our pruning experiments we used SGD optimizer.
+
+
+#### Export pruning algorithm description:
+This algorithm transforms trained with `Filter pruning algorithm` model with masked pruning (when convolution size remains
```
Comment (AlexKoff88) [inline docs/compression_algorithms/Pruning.md:94]: Propose changing ```trained with `Filter pruning algorithm` model with masked pruning``` to ```model trained with `Filter pruning` algorithm (...```

Comment (AlexKoff88) [reply docs/compression_algorithms/Pruning.md:94]: `remains same` -> `is still the same`

File: docs/compression_algorithms/Pruning.md [hunk 6]
```diff
@@ -83,6 +84,30 @@ Where ![a, k, b](https://microsoft.codecogs.com/gif.latex?a%2C%20k%2C%20b) - par
...
+ same but some filters are zeroed by mask) to smaller model with actually pruned operations.
+
+The algorithm consists of 3 stages:
+1. On the first stage, channel-wise masks from pruned operations propagate through the graph. Every operation has a function that can calculate the output mask by its attributes and the input masks. These calculated masks determine how the operations will be pruned.
```
Comment (AlexKoff88) [inline docs/compression_algorithms/Pruning.md:98]: `propagate` -> `is propagated`

File: docs/compression_algorithms/Pruning.md [hunk 7]
```diff
@@ -83,6 +84,30 @@ Where ![a, k, b](https://microsoft.codecogs.com/gif.latex?a%2C%20k%2C%20b) - par
...
+ There are three groups of operations:
+ - Operations that just pass the mask further unchanged. These operations include all activations and elementwise operations with one input, Batch Norms, Pullings.
```
Comment (AlexKoff88) [inline docs/compression_algorithms/Pruning.md:100]: `Batch Norms, Pulling` -> `BatchNorm, Pooling`

File: docs/compression_algorithms/Pruning.md [hunk 8]
```diff
@@ -83,6 +84,30 @@ Where ![a, k, b](https://microsoft.codecogs.com/gif.latex?a%2C%20k%2C%20b) - par
...
+ - Operations that cannot work with pruned input and do not skip the mask further. These operations include reshapes, matmul, reduce with some operations.
```
Comment (AlexKoff88) [inline docs/compression_algorithms/Pruning.md:101]: Looks strange, especially the end of the sentence. Do you mean `Reshape, MatMul, ReduceMean`?

File: docs/compression_algorithms/Pruning.md [hunk 9]
```diff
@@ -83,6 +84,30 @@ Where ![a, k, b](https://microsoft.codecogs.com/gif.latex?a%2C%20k%2C%20b) - par
...
+ - Operations that transform mask somehow. These operations include Convolutions (passes on a mask for this convolution), Concat (concatenates input masks),
```
Comment (AlexKoff88) [inline docs/compression_algorithms/Pruning.md:102]: `somehow` -> `in some way`

File: docs/compression_algorithms/Pruning.md (hunk 7 above)
Comment (AlexKoff88) [reply docs/compression_algorithms/Pruning.md:100]: By `elementwise operations with one input`, do you mean `Elementwise with constant`?

File: docs/compression_algorithms/Pruning.md [hunk 10]
```diff
@@ -83,6 +84,30 @@ Where ![a, k, b](https://microsoft.codecogs.com/gif.latex?a%2C%20k%2C%20b) - par
...
+  Elementwise (currently not supported, but potentially can intersect masks).
+
+2. In the second stage, a decision is made about which parts of the network will be pruned in accordance with the masks and which will not.
```
Comment (AlexKoff88) [inline docs/compression_algorithms/Pruning.md:105]: Please remove ` and which will not`

File: docs/compression_algorithms/Pruning.md [hunk 11]
```diff
@@ -83,6 +84,30 @@ Where ![a, k, b](https://microsoft.codecogs.com/gif.latex?a%2C%20k%2C%20b) - par
...
+This is necessary since some operations do not accept the pruned input and the path leading to it cannot be pruned.
+To do this, the special `can_prune` attribute is propagated through the network. For every operation, with information about
+    whether inputs and outputs accept pruned inputs and can be pruned, algorithm marks whether this operation can be pruned or not.
+The result is a valid graph for the pruned and non-pruned parts: that is, there is no operation that does not accept a pruned input,
+into which such input will come.
```
Comment (AlexKoff88) [inline docs/compression_algorithms/Pruning.md:110]: `into which such input will come` seems also unnessassary.

File: docs/compression_algorithms/Pruning.md (hunk 7 above)
Comment (mkaglins) [reply docs/compression_algorithms/Pruning.md:100]: Yes, I am talking about Elementwise with constant or Elementwise op like Power or sin that is calculated on one tensor input.

File: nncf/pruning/filter_pruning/algo.py [hunk 12]
```diff
@@ -147,9 +148,69 @@ def _apply_binary_mask_to_module_weight_and_bias(module, mask, module_name=""):
...
+            drow["Layer PR"] = 1 - new_modules_info[layer]['params_count'] / old_modules_info[layer]['params_count']
+            row = [drow[h] for h in header]
+            data.append(row)
+        table.add_rows(data)
+        return table
+
+    # pylint: disable=protected-access
     def export_model(self, filename, *args, **kwargs):
         """
         This function saving model without actually pruning the layers, just nullifies the necessary filters by mask.
```
Comment (vanyalzr) [inline nncf/pruning/filter_pruning/algo.py:193]: @mkaglins The description seems out of date. The model is actually pruned with this diff.
